CONTRIBUTING.rst

```diff
@@ -0,0 +1,292 @@
+  Both are acceptable. We weigh the cost of extending the scope of our current
+  work against the impact of perpetuating a problem. If we decide to make the
+  section compliant, we do so in a separate commit. That commit should not
+  introduce semantic changes and it should precede the commit that resovles the
+  issue.
+
+Type hints
+**********
+
+* We use type hints both to document intent and to facilitate type checking by
+  the IDE as well as additional tooling.
+  
+* When defining type hints for a function or method, we do so for all its
+  parameters and return values.
+  
+* We prefer the generic types from ``typing`` over those from the
+  ``collections`` module e.g., ``Mapping[K,V]`` over ``dict``.
+
+* Due to the prominence of JSON in the project we annotate variables containing
+  deserialized JSON as such, using the ``JSON`` type from ``azul.typing``.
+  
+
+
+Version Control
+===============
+
+* Feature branches are integrated by rebasing or squashing. We only use merges
+  between deployment branches, either to promote changes in their natural
+  progression from one deployment to the next or to
+
+* We commit independent changes separately. If two changes could be applied in
+  either order, they should occur in separate commits. Two changes A and B of
+  which B depends on A may still be comitted separately if B represents an
+  extension of A that we might want to revert while leaving A in place.
+
+* In theory, every individual commit should pass unit and integration tests. In
+  practice, on PR branches with long histories not intented to be squashed, not
+  every commit is built in CI. This is acceptable.
+
+  .. note::
+
+    I am not a fan the above rule but the desire to maintain a linear history
+    by rebasing PR branches as opposed to merging them demands that loophole.
+    When pushing a rebased PR branch, we'd have to build every commit on that
+    branch individually. The loophole can be avoided by creating narrowly
+    focused PRs with only one logical change and therefore one commit.
+    Therefore, we consider the creation of PRs with longer histories to be a
+    privilege of trusted, long-time contributors. See ``
+
+
+* If a commit resolves (or contributes to the resolution of) an issue, we
+  mention that issue at the end of the commit title::
+
+    Reticulate them splines for good measure (#123)
+
+  Note that we don't use Github resolution keywords like "fixes" or "resolves".
+  Any mention of those preceding an issue reference in a title would
+  automatically close the issue as soon as the commit lands on the default
+  branch. This is undesirable as we want to continue to track issues in
+  Zenhub's *Merged* and *Done* pipelines even after the commit lands on
+  `develop`.
+
+* We value `expressive and concise commit message titles`_ and we use Github's
+  limit on the length of a commit message title. A commit title is too long if
+  (and only if) Github truncates or wraps the title in its UI.
```

Comment:
I think their limit is 50 characters (that's the same as what's recommended in your commit message link.)

Reply:
GitHub shortens titles longer than 72 by truncating it at 69 appending three dots. 50 is too short IMO so I'll make it 72. I disagree with the Linux kernel repo limit of 50 which even Linux devs don't stick with (https://preslav.me/2015/02/21/what-s-with-the-50-72-rule/). Since GH uses a proportional font, even titles shorter than 73 may display as wrapped over two lines, titles with lots of wide Ws or Ms, for example. I think that's acceptable given that its improbable in English phrases no longer than 72 characters.